### PR TITLE
Detect Network Instance IP subnet conflict

### PR DESCRIPTION
**Pull request by milan-zededa**
Zedrouter is able to detect if NI IP subnet overlaps with the subnet of another NI or with a device port network. When a conflict is detected for a new NI, it is blocked from being created. If IP conflict arises for an already created NI (e.g. a device port later received IP from an overlapping subnet), the NI is unconfigured and VIFs are set DOWN (not deleted). In both cases an error is reported for the NI.

This is crucial because, previously, a NI with a subnet overlapping with a device port network would result in the device permanently losing controller connectivity (due to routing conflicts), with very limited options to restore connectivity and make the device manageable again. Instead, the device should remain online, inform the user about the issue, and allow to correct the IP configuration.

When IP conflict is detected for an already created NI with an app connected to it, we intentionally do not halt and undeploy the app. Instead, the app is left running, just VIFs connected to the problematic NI lose their connectivity. This allows the user to fix the network config or at least backup the application data when IP conflict arises.

To enable the deletion of NI when an IP conflict is detected, followed by its later recreation when the conflict is resolved, and to reconnect already running apps, few enhancements had to be made to the NI Reconciler, particularly in the area of VIF bridging.

### CODE REVIEW

**Comment by milan-zededa** [comment]
Note that I will be backporting this to 11.0 and 10.4 (in 9.4 the zedrouter is vastly different and would require to make the patch from the ground up).

File: pkg/pillar/cmd/zedrouter/zedrouter.go [hunk 1]
```diff
@@ -244,22 +244,22 @@ func (z *zedrouter) init() (err error) {
 
 	z.peUsagePersist, err = persistcache.New(types.PersistCachePatchEnvelopesUsage)
 	if err != nil {
-		log.Fatal(err)
+		return err
```
Comment (eriknordmark) [inline pkg/pillar/cmd/zedrouter/zedrouter.go:247]: Is this a separate fix (which would make sense in a separate commit)?

**Comment by eriknordmark** [comment]
> To enable the deletion of NI when an IP conflict is detected, followed by its later recreation when the conflict is resolved, and to reconnect already running apps, few enhancements had to be made to the NI Reconciler, particularly in the area of VIF bridging.

Is this the reason for the introduction of the BridgePort and the various BridgeIfName changes?

**Review by eriknordmark — APPROVED** [review]
Run tests

File: pkg/pillar/cmd/zedrouter/validation.go [hunk 2]
```diff
@@ -33,63 +33,32 @@ func (z *zedrouter) doNetworkInstanceSanityCheck(
...
-	return false, nil
+	return nil
 }
 
 func (z *zedrouter) doNetworkInstanceSubnetSanityCheck(
-	config *types.NetworkInstanceConfig) (niConflict bool, err error) {
+	config *types.NetworkInstanceConfig) error {
 	if config.Subnet.IP == nil || config.Subnet.IP.IsUnspecified() {
-		return false, fmt.Errorf("subnet unspecified for %s-%s: %+v",
+		return fmt.Errorf("subnet unspecified for %s-%s: %+v",
```
[the remark this answers is not in the document]
Comment (milan-zededa) [reply pkg/pillar/cmd/zedrouter/validation.go:58]: This function `doNetworkInstanceSubnetSanityCheck` is called only if `config.IpType` is not `AddressTypeNone`, so it will not be called for switch network instance.
Note that IP conflict detection was now moved to a separate function: `checkNetworkInstanceIPConflicts`.

File: pkg/pillar/cmd/zedrouter/zedrouter.go (hunk 1 above)
Comment (milan-zededa) [reply pkg/pillar/cmd/zedrouter/zedrouter.go:247]: Yes, it is a minor unrelated fix, I will move it to a separate commit.

Comment (milan-zededa) [reply pkg/pillar/cmd/zedrouter/zedrouter.go:247]: Done

**Comment by christoph-zededa** [comment]
Instead of detecting the subnet overlap, wouldn't it also be possible to bind the socket to the interface (`syscall.SetsockoptString(int(fd), syscall.SOL_SOCKET, syscall.SO_BINDTODEVICE, device)`) when sending requests?

**Comment by eriknordmark** [comment]
> Instead of detecting the subnet overlap, wouldn't it also be possible to bind the socket to the interface (`syscall.SetsockoptString(int(fd), syscall.SOL_SOCKET, syscall.SO_BINDTODEVICE, device)`) when sending requests?

@christoph-zededa 
The EVE management traffic already does that, but using a different mechanism (binding to the IP address of the interface and then using an IP rule based on that source IP address to select the routing table, and that routing table contains the routes out that interface.)
But that is insufficient when there are some IP subnet conflicts.

**Comment by milan-zededa** [comment]
@christoph-zededa  @eriknordmark Note that I plan to investigate options how to isolate network instances from uplinks and allow even overlapping subnets. In the past, I tested VRFs and network namespaces. Both work but introduce quite some complexity and data-plane overhead. Alternative solution could be to rework policy-based routing and to experiment with `SO_BINDTODEVICE` as you suggested. The problem with interface binding is that it does not cover all cases (e.g. not only pillar uses the mgmt IP) and app traffic from the conflicting NI will still have broken routing. In any case, I would need more time to research all the options and plan to do it in the next development cycle. This PR is a quick patch just to prevent device loosing connectivity (and all other problems that subnet overlap creates). Simple enough so that it can be backported to LTS versions.

**Comment by christoph-zededa** [comment]
> @christoph-zededa @eriknordmark Note that I plan to investigate options how to isolate network instances from uplinks and allow even overlapping subnets. In the past, I tested VRFs and network namespaces. Both work but introduce quite some complexity and data-plane overhead. Alternative solution could be to rework policy-based routing and to experiment with `SO_BINDTODEVICE` as you suggested. The problem with interface binding is that it does not cover all cases (e.g. not only pillar uses the mgmt IP) and app traffic from the conflicting NI will still have broken routing. In any case, I would need more time to research all the options and plan to do it in the next development cycle. This PR is a quick patch just to prevent device loosing connectivity (and all other smaller problems that subnet overlap creates). Simple enough so that it can be backported to LTS versions.

Thank you for the explanation - I see it would take some more time to implement.
